src/rpc/client.zig

```diff
@@ -1,2338 +1,537 @@
-        var optionsObj = std.StringArrayHashMap(json.Value).init(self.allocator);
-        defer optionsObj.deinit();
-        try optionsObj.put("encoding", .{ .string = options.encoding.string() });
-        try optionsObj.put("commitment", .{ .string = self.defaultCommitmentOr(options.commitment).string() });
-        try optionsObj.put("withContext", .{ .bool = true });
+    pub fn getLeaderSchedule(self: *Client, allocator: std.mem.Allocator, maybe_epoch: ?Epoch, config: GetLeaderScheduleConfig) !Result(types.LeaderSchedule) {
+        var arena = std.heap.ArenaAllocator.init(allocator);
+        errdefer arena.deinit();
 
-        if (options.minContextSlot) |minContextSlot| {
-            try optionsObj.put("minContextSlot", .{ .integer = @as(i64, @intCast(minContextSlot)) });
-        }
+        var params_builder = Request.ParamsBuilder.init(arena.allocator());
+        try params_builder.addOptionalArgument("{d}", maybe_epoch);
+        try params_builder.addConfig(config);
 
-        if (options.dataSlice) |dataSlice| {
-            var dataSliceObj = std.StringArrayHashMap(json.Value).init(self.allocator);
-            defer dataSliceObj.deinit();
-            try dataSliceObj.put("length", .{ .integer = @as(i64, @intCast(dataSlice.length)) });
-            try dataSliceObj.put("offset", .{ .integer = @as(i64, @intCast(dataSlice.offset)) });
-            try optionsObj.put("dataSlice", .{ .object = dataSliceObj });
-        }
+        const leader_schedule_json = try self.sendFetchRequest(arena.allocator(), std.json.Value, .{
+            .method = "getLeaderSchedule",
+            .params = try params_builder.build(),
+        });
 
-        if (options.filters) |filters| {
-            var filtersList = std.ArrayList(json.Value).init(self.allocator);
-            defer filtersList.deinit();
-
-            for (filters) |filter| {
-                var filterObj = std.StringArrayHashMap(json.Value).init(self.allocator);
-                defer filterObj.deinit();
-                if (filter.memcmp) |memcmp| {
-                    var memcmpObj = std.StringArrayHashMap(json.Value).init(self.allocator);
-                    defer memcmpObj.deinit();
-                    try memcmpObj.put("bytes", .{ .string = memcmp.bytes });
-                    try memcmpObj.put("offset", .{ .integer = @as(i64, @intCast(memcmp.offset)) });
-                    try memcmpObj.put("encoding", .{ .string = memcmp.encoding.string() });
-
-                    try filterObj.put("memcmp", .{ .object = filterObj });
-                } else if (filter.dataSize) |dataSize| {
-                    try filterObj.put("dataSize", .{ .integer = @as(i64, @intCast(dataSize)) });
-                }
-
-                try filtersList.append(.{ .object = filterObj });
+        var leader_schedule = types.LeaderSchedule.init(arena.allocator());
+        var json_iter = leader_schedule_json.object.iterator();
```

Comment:
Sort of general comment/question about accessing these union fields directly: is there any data sanitization at any point in the pipeline to prevent this from triggering (checked) illegal behavior?

Reply:
4d408cd